debian/fina.8
.\"                                      Hey, EMACS: -*- nroff -*-
.\" First parameter, NAME, should be all caps
.\" Second parameter, SECTION, should be 1-8, maybe w/ subsection
.\" other parameters are allowed: see man(7), man(1)
.TH FINA 8 "February 26, 2008"
.\" Please adjust this date whenever revising the manpage.
.\"
.\" Some roff macros, for reference:
.\" .nh        disable hyphenation
.\" .hy        enable hyphenation
.\" .ad l      left justify
.\" .ad b      justify to both left and right margins
.\" .nf        disable filling
.\" .fi        enable filling
.\" .br        insert line break
.\" .sp <n>    insert n+1 empty lines
.\" for manpage-specific macros, see man(7)
.SH NAME
fina \- The Netfilter script for admins who know what they're doing.
.SH SYNOPSIS
.B fina
.RI [ options ] 
.br
.B fina
.RI  --help|-h
.SH DESCRIPTION
This manual page documents how \fBfina\fP works and why it does it the way it does.
If you need a quick start in setting up \fBfina\fP, look at the file INSTALL
that came with the package.
.SH OPTIONS
\fBFina\fP follows follow the usual GNU command line syntax, with long
options starting with two dashes (`-').
.TP
.B \-h, \-\-help
Show summary of options.
.TP
.B \-v, \-\-version
Show version of program.
.TP
.B \-p, \-\-pretend
Don't load rules, just print them to stdout. (default)
.TP
.B \-l, \-\-load
Create rules and load them (opposition to -p)
.TP
.B \-m, \-\-minimal
Load minimal ruleset (implies -l)
.TP
.B \-t, \-\-terse
Remove all comments and empty lines from output when in 'pretend' mode
.TP
.B \-k, \-\-keep
Keep files generated in /tmp
.PP
The modes 
.B -p 
and 
.B -l 
are mutually exclusive, the last of those encountered
on the command line is used.

.SH DESIGN
\fBFina\fP is the result of working with various attempts at wrapping iptables
(rather: Netfilter) with scripts and other infrastructure in order to
make daily administration of rules easier. 

Many admins need a script that can easily be automated
remotely. \fBFina\fP was designed with that goal in mind

Another goal for \fBfina\fP  is that it has to be secure by
default - without getting into the admins way too much. That is to say,
\fBfina\fP only creates rules that the admin expressly configures. Of course, a
sensible set of configuration 
.I examples 
is provided.

The third hard requirement is that the script should be capable of doing
everything Netfilter can. That means that no matter how complicated the
NAT and prerouting magic is, the script must be able to handle
it. One side effect of this is that it must handle huge filter sets at
least as easily as Netfilter itself.

The fourth requirement is a bit complicated. Sometimes, it may be
necessary to change variables in /proc after modules have been loaded
but before rules referring to those modules are added. Also, sometimes
it might be necessary, to do stuff after the packet filter has been
loaded. To this end, \fBfina\fP may be used to run two scripts called
"pre-up.sh" and "post-up.sh". If those files exist and are executable,
they are executed at the corresponding times.

Finally, the script should be lightweight, having as few external
dependencies as possible. Also, the script itself should be as simple
and small as possible. This is because simple things have fewer ways in
which they can fail.

.SH OPERATION

Due to the four hard requirements mentioned above, the script doesn't
actually "know" all that much about Netfilter. This way, it easily
incorporates everything Netfilter can do - without having a moving
target of specs to adhere to.

Basically, \fBfina\fP assembles an iptables-restore compatible rule dump from
snippets configured/edited by the admin of the machine. It then tries to
load the whole set. If this fails in any way, the old rule set is
restored.

Naturally, this means that \fBfina\fP does next to nothing for the admin
when it comes to 
.I generating
rules on the machine. This is a conscious
decision. For one, whoever uses/configures a packet filter should be
aware of all the consequences of his actions. This includes for example
knowing why it is bad to drop all ICMP traffic.

Another reason is that providing examples or even recipes for packet
filters is something 
.I documentation
should do. It's not a good idea to
have a black box do "simply the right thing". Usually, there is a
non-trivial amount of users for which The Right Thing can't be easily
guessed.

.SH STARTING AND STOPPING

If \fBfina\fP is started via the init script, one critical extra step is
taken: the script tries to load /etc/fina/minimal.rules with
iptables-restore. This comes in handy if you've updated the kernel and
forgotten parts of the iptables modules - your nice big config file
probably won't load and your machine is either vulnerable or
inaccessible to you. Usually, you'll want to have a very simple rule set
in here (without conntracking) that allows access from your management
IPs. You can load this ruleset using the -m commandline switch. This can
be handy in emergency situations (think of it as a PANIC button).

\fBFina\fP expects one configuration file, /etc/fina/fina.cfg. This is
actually simply a shell script that is sourced from the main \fBfina\fP
script. It contains (by default) one variable that specify one 
additional location, that of the rule directory. First, \fBfina\fP checks if
/etc/fina/pre-up.sh exists and is executable and if so runs it. This is
the place to load modules or change stuff in proc if the need arises.

The second location is more important. It specifies the directory which
contains the rule snippets that \fBfina\fP should assemble. Usually, this is
located at /etc/fina/rules.d/. \fBFina\fP then proceeds to load all files from
said directory (and its sub-directories) that end in .rules. In order to
have reliable order when assembling, the files are usually prefixed with
two- or three-digit numbers. 

The expected format of the files is of no concern to \fBfina\fP. \fBFina\fP itself
will only assemble them in order into one file. After this, it will make
a backup of the current rule set (using iptables-save) and put it in a
location specified in /etc/fina/fina.cfg. Then, \fBfina\fP will try to load
the generated file using iptables-restore. If this fails for whatever
reason, it will try to load the old rule set and exhibit an appropriate
error message plus whatever message it got from the failing
iptables-restore command. 

Finally, \fBfina\fP will execute /etc/fina/post-up.sh if it exists and is
executable.

.SH EXAMPLES

Examples for the configuration of \fBfina\fP located at /var/lib/fina.

.SH SEE ALSO
.BR iptables (8), 
.BR iptables-save (8).
.BR iptables-restore (8).
.SH AUTHOR
\fBfina\fP (and this manpage) was written by Tobias Klausmann
.PP
This manual page is maintained by Alexander Gedak.


debian/changelog
fina (0.1.4-3) unstable; urgency=low

  * 

 -- root <root@debian.mk.freenet-rz.de>  Wed, 27 Feb 2008 15:27:14 +0100

fina (0.1.4-2) unstable; urgency=low

  * Nearly complete 

 -- Alexander Gedak <packages@freenet-rz.de>  Wed, 27 Feb 2008 14:20:34 +0100

fina  (0.1.4-1)  unstable; urgency=low

  * Initial release

 -- Alexander Gedak <packages@freenet-rz.de>  Tue, 26 Feb 2008 17:41:22 +0100



fina.sh
#!/bin/bash

# $Id: fina.sh 22 2007-11-18 16:16:45Z klausman $

# Fina v0.1 Copyright (C) 2007 Tobias Klausmann
# Released under the GPLv2. See the file COPYING for details.

# This is the place where the main config file resides.
FINACFG="/etc/fina/fina.cfg"

# If on your system, some tools are on a different PATH, you can
# configure them here
FIND=$(which find)
SORT=$(which sort)
MODPROBE=$(which modprobe)
IPTRESTORE=$(which iptables-restore)
IPTSAVE=$(which iptables-save)
# --------------- Nothing to configure below here --------------- #

PROGNAME="$0"
VERSION="0.1.4"
DEBUG="0"

# Help function
function print_help() {
    {
    echo "Usage: ${PROGNAME} --help|-h"
    echo "   or: ${PROGNAME} [OPTIONS]" 
    echo "" 
    echo "Options:" 
    echo "   -v, --version    Output script version, then exit."
    echo "   -p, --pretend    Don't load rules, just print them to stdout. (default)" 
    echo "   -l, --load       Create rules and load them" 
    echo "   -m, --minimal    Load minimal ruleset (implies -l)"
    echo "   -t, --terse      Remove all comments and empty lines from output when in"
    echo "                    'pretend' mode"
    echo "   -k, --keep       Keep files generated in /tmp"
    echo "" 
    echo "The modes -p and -l are mutually exclusive, the last one of those encountered"
    echo "on the command line is used."
    } >&2
}

function print_version() {
    echo "Fina version $VERSION"
    echo "(C) 2007 Tobias Klausmann." 
    echo "This is free software"  
    echo "You may redistribute copies of it under the terms of the GNU GPLv2"
    echo "http://www.gnu.org/licenses/old-licenses/gpl-2.0.html"
    echo "There is NO WARRANTY, to the extent permitted by law."
}

# Debug printing 
function dprint() {
    if [[ $DEBUG == 1 ]]; then
        echo  -e "$@" >&2
    fi
}

# Error printing 
function eprint() {
    echo  -e "$@" >&2
}

# Exit with message and error code
function die() {
    eprint "$@"
    exit -1
}

# Function that retrieves all rule files in lexical order
function get_rule_files () {
    # The trailing / is necessary in case ${RULESDIR} is a symlink
    $FIND ${RULESDIR}/ -name \*.rules -a \( -type f -o -type l \) -print|$SORT
}

# Function to create a temporary space for the file(s) we create
function get_tmpdir () {
    local UMASK=$(umask) # So we can restore it
    local OK=1
    local TRIES=3
    local TMPDIR=''
    umask 0077
    while [[ $OK != 0 && "$TRIES" -gt 0 ]]; do
        TMPDIR=/tmp/fina-$RANDOM$RANDOM
        TRIES=$(($TRIES-1))
        dprint "Trying to create $TMPDIR (tries left: $TRIES)" >&2
        mkdir "$TMPDIR" && OK=0
    done
    if [[ $OK != 0 ]]; then
        echo "Could not create TMPDIR."
        exit -1
    fi
    echo "${TMPDIR}"
    umask "$UMASK"
}

# Defaults
PRETEND=1
TERSE=0
KEEPTEMP=0
LOAD=0
MINIMAL=0

# Get cfg file contents
. "${FINACFG}" || die "Could not load config from $FINACFG"

# Parse cmdline
while test -n "$1"; do
    case "$1" in
        --help|-h)
            print_help
            exit 0
            ;;
        --version|-v)
            print_version
            exit 0
            ;;
        --pretend|-p)
            PRETEND=1
            ;;
        --keep|-k)
            KEEPTEMP=1
            ;;
        --load|-l)
            PRETEND=0
            ;;
        --terse|-t)
            TERSE=1
            ;;
        --minimal|-m)
            MINIMAL=1
            ;;
        *)
            echo "Unknown option '$1'"
            print_help
            exit 1
    esac
    shift
done

dprint "After cmdline. Programs are:"
dprint "FIND='$FIND'"
dprint "SORT='$SORT'"
dprint "MODPROBE='$MODPROBE'"
dprint "IPTRESTORE='$IPTRESTORE'"
dprint "IPTSAVE='$IPTSAVE'"
dprint ""
dprint "Program mode: PRETEND=$PRETEND KEEPTEMP=$KEEPTEMP LOAD=$LOAD MINIMAL=$MINIMAL"

# If we are asked to load the minimal ruleset, things are rather simple
if [[ $MINIMAL == 1 ]]; then
    if [[ -r /etc/fina/minimal.rules ]]; then
        echo -n 'Trying to load minimal rule set...'
        # if this fails, theres nothing we can do except print an errmsg
        MSG=$(iptables-restore < /etc/fina/minimal.rules 2>&1)
        RETVAL=$?
        if [[ $RETVAL != 0 ]]; then
            echo "failed ($MSG)" >&2
            exit $RETVAL
        else
            echo "ok"
            exit 0
        fi
    fi
fi

# First, execute pre-up.sh
if [[ -x /etc/fina/pre-up.sh ]]; then
    /etc/fina/pre-up.sh
fi

# Now, lets get the rule files
dprint "Searching rules in '${RULESDIR}'"
dprint "Found:\n$(get_rule_files)"

# Get a tmpfile to put assembled rules in
TMPDIR=$(get_tmpdir)

# Save the old rules if we're not pretending -or- if we're told to keep stuff
if [[ $PRETEND != 1 || $KEEPTEMP == 1 ]]; then
    $IPTSAVE > $TMPDIR/rules.old
fi


# Generate the new rule file
{
    echo -n "#Fina# Generated by Fina v${VERSION} on "
    date -Is
    echo 
    # We can't re-use $RULEFILES here since we need to handle spaces 
    # in filenames
    get_rule_files|while read RULEFILE; do 
        echo "#Fina# BEGIN of file '$RULEFILE'"
        # Comments only work if there is *nothing* before the #, so
        # we remove all trailing comments (first expression) and remove
        # all whitespace-sequences before comment line (second expression)
        sed -re 's/([^	 ]+)[	 ]*#.*/\1/g' -e 's/^[   ]*#/#/g' "$RULEFILE"
        RETVAL=$?
        if [[ $RETVAL != 0 ]]; then
            eprint "Could not open '$RULEFILE' (Error code '$RETVAL')"
            eprint "I have left all the files I generated in $TMPDIR"
            exit -2
        fi
        echo "#Fina# END of file '$RULEFILE'"
    done
    echo -n "#Fina# Generation complete on "
    date -Is 
} > "$TMPDIR/rules.new"

if [[ $PRETEND == 1 ]]; then
    if [[ $TERSE == 1 ]]; then
        grep -Ev '(^$|^[[:space:]]*#)' -- "$TMPDIR/rules.new"
    else
        cat "$TMPDIR/rules.new"
    fi
else
    dprint "Loading rules from file $TMPDIR/rules.new"
    MESG=$(iptables-restore < "$TMPDIR/rules.new" 2>&1)
    RETVAL=$?
    if [[ $RETVAL != 0 ]]; then
        eprint "Something went wrong (retval $RETVAL). Message:"
        eprint "$MESG"
        eprint "Loading old ruleset from $TMPDIR/rules.old"
        MESG=$(iptables-restore < "$TMPDIR/rules.old" 2>&1)
        RETVAL=$?
        if [[ $RETVAL != 0 ]]; then
            eprint "Even the old ruleset could not be loaded. Message:"
            eprint "$MESG"
        fi
        eprint "I have left all the files I generated in $TMPDIR"
        exit -1
    fi
fi

if [[ $KEEPTEMP != 1 ]]; then
    rm -rf $TMPDIR
fi

# Finally, execute post-up.sh
if [[ -x /etc/fina/post-up.sh ]]; then
    /etc/fina/post-up.sh
fi

# vim: tabstop=4:shiftwidth=4:smarttab:expandtab:softtabstop=4:smartindent
